File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable
from typing import List, Optional, Generator

def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts
    
# Write your first then lower case function here
def first_then_lower_case(a: List[str], p: Callable[[str], bool]) -> Optional[str]:
    for item in a:
        if p(item):
            return item.lower()
    return None

# Write your powers generator here
def powers_generator(*, base: int, limit: int) -> Generator[int, None, None]:
    value = 1
    while value <= limit:
        yield value
        value *= base

# Write your say function here
def say(word: str = None) -> Callable[[str], str] | str:
    words = []
    def inner(next_word: str = None) -> Callable[[str], str] | str:
        nonlocal words
        if next_word is None:
            return " ".join(words)
        words.append(next_word)
        return inner

    if word is not None:
        words.append(word)

    if word == "" and len(words) == 1:
        return inner

    return inner() if word is None else inner

# Write your line count function here
def meaningful_line_count(filename: str) -> int:
    try:
        with open(filename, 'r') as file:
            count = 0
            for line in file:
                stripped_line = line.strip()
                if stripped_line and not stripped_line.startswith("#"):
                    count += 1
            return count
    except FileNotFoundError as e:
        raise FileNotFoundError(f"No such file: {filename}") from e

# Write your Quaternion class here
@dataclass(frozen=True)
class Quaternion:
    a: float
    b: float
    c: float
    d: float

    @property
    def coefficients(self):
        return (self.a, self.b, self.c, self.d)

    @property
    def conjugate(self):
        return Quaternion(self.a, -self.b, -self.c, -self.d)

    def __add__(self, other):
        return Quaternion(
            self.a + other.a,
            self.b + other.b,
            self.c + other.c,
            self.d + other.d
        )

    def __mul__(self, other):
        a = self.a * other.a - self.b * other.b - self.c * other.c - self.d * other.d
        b = self.a * other.b + self.b * other.a + self.c * other.d - self.d * other.c
        c = self.a * other.c - self.b * other.d + self.c * other.a + self.d * other.b
        d = self.a * other.d + self.b * other.c - self.c * other.b + self.d * other.a
        return Quaternion(a, b, c, d)

    def __eq__(self, other):
        return (self.a, self.b, self.c, self.d) == (other.a, other.b, other.c, other.d)

    def __str__(self):
        components = []

        if self.a != 0:
            components.append(f"{self.a}")

        if self.b != 0:
            if self.b == 1:
                components.append("+i" if components else "i")
            elif self.b == -1:
                components.append("-i")
            else:
                components.append(f"{'+' if self.b > 0 and components else ''}{self.b}i")

        if self.c != 0:
            if self.c == 1:
                components.append("+j" if components else "j")
            elif self.c == -1:
                components.append("-j")
            else:
                components.append(f"{'+' if self.c > 0 and components else ''}{self.c}j")

        if self.d != 0:
            if self.d == 1:
                components.append("+k" if components else "k")
            elif self.d == -1:
                components.append("-k")
            else:
                components.append(f"{'+' if self.d > 0 and components else ''}{self.d}k")

        return "".join(components) or "0"

File: lua/exercises.lua
function change(amount)
  if math.type(amount) ~= "integer" then
    error("Amount must be an integer")
  end
  if amount < 0 then
    error("Amount cannot be negative")
  end
  local counts, remaining = {}, amount
  for _, denomination in ipairs({25, 10, 5, 1}) do
    counts[denomination] = remaining // denomination
    remaining = remaining % denomination
  end
  return counts
end  

-- Write your first then lower case function here
function first_then_lower_case(a, p)
  for _, item in ipairs(a) do
    if p(item) then
      return string.lower(item)
    end
  end
  return nil
end

-- Write your powers generator here
function powers_generator(base, limit)
  local generator = function()
      local power = 0
      local value = base ^ power

      while value <= limit do
          coroutine.yield(value)
          power = power + 1
          value = base ^ power
      end
  end
  return coroutine.create(generator)
end

-- Write your say function here
function say(word)
  local words = {}
  
  local function chain(new_word)
      if new_word == nil then
          return table.concat(words, " ")
      end
      table.insert(words, new_word)
      return chain
  end

  return chain(word)
end

-- Write your line count function here
function meaningful_line_count(filename)
  local file, err = io.open(filename, "r")
  if not file then
      error("No such file")
  end
  
  local count = 0
  for line in file:lines() do
      local trimmed_line = line:match("^%s*(.-)%s*$")
      if #trimmed_line > 0 and trimmed_line:sub(1, 1) ~= "#" then
          count = count + 1
      end
  end
  
  file:close()
  return count
end

-- Write your Quaternion table here
Quaternion = {}
Quaternion.__index = Quaternion

function Quaternion.new(a, b, c, d)
  local self = setmetatable({}, Quaternion)
  self.a = a
  self.b = b
  self.c = c
  self.d = d
  return self
end

function Quaternion.__add(q1, q2)
  return Quaternion.new(q1.a + q2.a, q1.b + q2.b, q1.c + q2.c, q1.d + q2.d)
end

function Quaternion.__mul(q1, q2)
  return Quaternion.new(
      q1.a * q2.a - q1.b * q2.b - q1.c * q2.c - q1.d * q2.d,
      q1.a * q2.b + q1.b * q2.a + q1.c * q2.d - q1.d * q2.c,
      q1.a * q2.c - q1.b * q2.d + q1.c * q2.a + q1.d * q2.b,
      q1.a * q2.d + q1.b * q2.c - q1.c * q2.b + q1.d * q2.a
  )
end

function Quaternion:coefficients()
  return {self.a, self.b, self.c, self.d}
end

function Quaternion.__eq(q1, q2)
  return q1.a == q2.a and q1.b == q2.b and q1.c == q2.c and q1.d == q2.d
end

function Quaternion:conjugate()
  return Quaternion.new(self.a, -self.b, -self.c, -self.d)
end

function Quaternion:__tostring()
  local output = ""
  local terms = {"", "i", "j", "k"}
  local coefficients = self:coefficients()

  for i, coefficient in ipairs(coefficients) do
      if coefficient ~= 0 then
          if output ~= "" and coefficient > 0 then
              output = output .. "+"
          end
          
          local abs_val = math.abs(coefficient)

          if i == 1 then
              output = output .. tostring(coefficient)
          else
              if abs_val == 1 then
                  output = output .. (coefficient < 0 and "-" or "") .. terms[i]
              else
                  output = output .. tostring(coefficient) .. terms[i]
              end
          end
      end
  end
  return output == "" and "0" or output
end




